java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -959,53 +964,67 @@ class DatabaseImpl(
         schema: Schema,
         db: SQLiteDatabase,
         counters: Counters? = null
-    ): TypeId = schemaMutex.withLock {
-        schemaTypeMap[schema.hash]?.let {
-            counters?.increment(DatabaseCounters.ENTITY_SCHEMA_CACHE_HIT)
-            return@withLock it
-        }
-        counters?.increment(DatabaseCounters.ENTITY_SCHEMA_CACHE_MISS)
-
-        return db.transaction {
-            counters?.increment(DatabaseCounters.INSERT_ENTITY_TYPE_ID)
+    ): TypeId = db.transaction {
+        var cacheHit = false
+        val schemaTypeId = schemaTypeMap.updateAndGet { currentMap ->
+            currentMap[schema.hash]?.let {
+                cacheHit = true
+                return@updateAndGet currentMap
+            }
+            cacheHit = false
             val content = ContentValues().apply {
                 put("name", schema.hash)
                 put("is_primitive", false)
             }
-            val schemaTypeId = insertOrThrow(TABLE_TYPES, null, content)
+            insertWithOnConflict(
+                TABLE_TYPES,
+                null,
+                content,
+                SQLiteDatabase.CONFLICT_IGNORE
+            )
+            val id = rawQuery("SELECT id FROM types WHERE name = ?", arrayOf(schema.hash))
```

Comment:
now cacheHit is always false here. Is it possible for the schema to be in the database but not in the in-memory map? because in that case you would be inserting fields again.

Reply:
It's not possible, thanks to the coupling of the atomic update block and the fact that it's run within a db transaction.